modules/internal/ChapelIO.chpl

```diff
@@ -345,50 +349,225 @@ module ChapelIO {
+          // Any error reading superclass must be preserved.
+          superclass_error = reader.error();
         }
       }
   
+      //writeln("in readThisFieldsDefaultImpl after super. type is ", t:string, " needsComma=", needsComma);
+
       if !isUnionType(t) {
         // read all fields for classes and records
-  
-        for param i in 1..num_fields {
-          if isType(__primitive("field value by num", x, i)) ||
-             isParam(__primitive("field value by num", x, i)) {
-             // do nothing, don't read types or params
-          } else {
-            if !isBinary {
+        if isBinary {
+          for param i in 1..num_fields {
+            if isType(__primitive("field value by num", x, i)) ||
+               isParam(__primitive("field value by num", x, i)) {
+              // do nothing, don't read types or params
+            } else {
+              reader.readwrite(__primitive("field value by num", x, i));
+            }
+          }
+        } else if num_fields > 0 {
+          // track whether or not we've read all of the fields.
+
+          // this tuple helps us not read the same field twice.
+          var read_field:(num_fields)*bool;
+          var num_to_read = 0;
+          var num_read = 0;
+          for param i in 1..num_fields {
+            if isType(__primitive("field value by num", x, i)) ||
+               isParam(__primitive("field value by num", x, i)) {
+              // do nothing, don't read types or params
+            } else {
+              num_to_read += 1;
+            }
+          }
+
+          // the order should not matter.
+          while num_read < num_to_read {
+            //writeln("IN LOOP read_field=", read_field);
+
+            if needsComma {
+              // read a comma
+
+              //writeln("READING COMMA");
               var comma = new ioLiteral(",", true);
-              if !first then reader.readwrite(comma);
-    
-              var fname = new ioLiteral(__primitive("field num to name", t, i), true);
-              reader.readwrite(fname);
-    
-              var st = reader.styleElement(QIO_STYLE_ELEMENT_AGGREGATE);
-              var eq:ioLiteral;
-              if st == QIO_AGGREGATE_FORMAT_JSON {
-                eq = new ioLiteral(":", true);
+              reader.readwrite(comma);
+
+              if !reader.error() {
+                needsComma = false; // we read a comma
+              } else if reader.error() == EFORMAT {
+                //writeln("NO COMMA");
+                // break out of the loop if we didn't read a comma
+                // and we're expecting to read one.
+                break;
+              }
+            }
+
+            // find a field name that matches.
+            // TODO: this is not particularly efficient. If we
+            // have a lot of fields, this is O(n**2), and there
+            // are other potential problems with string reallocation.
```

Comment:
The use cases I've been working on have a small number of fields, so this is unlikely to be a performance problem there. I could add the regexp version but it would add complexity and make the feature dependent on having Chapel built with regexp support. I plan to leave it as-is for those reasons.